### PR TITLE
Alternative solvent extension for the black oil model.

### PR DESCRIPTION
**Pull request by osae**
A first version, depending on corresponding pr for opm-models (extboModels) and opm-common (kwPVTEXTBO).

### CODE REVIEW

File: ebos/eclfluxmodule.hh
Comment (totto82) [inline ebos/eclfluxmodule.hh:277]: Maybe an explanation why this is needed?

File: ebos/ecloutputblackoilmodule.hh
Comment (totto82) [inline ebos/ecloutputblackoilmodule.hh:688]: This seems a bit hardcoded. The reference density should be available here.

File: ebos/ecloutputblackoilmodule.hh
Comment (totto82) [inline ebos/ecloutputblackoilmodule.hh:1088]: Opm::data::TargetType::RESTART_AUXILIARY would require the user to set the flag --enable-opm-rst-file=true
I don't know the policy of the RESTART_AUXILIARY and RESTART_SOLUTION but maybe it is better to use the RESTART_SOLUTION here since the extended bo model is specific to opm flow.

File: opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp
Comment (totto82) [inline opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp:1236]: Why not use the perfRateSolvent_ ? That will cleanup the code IMO.

**Review by totto82 — CHANGES_REQUESTED** [review]
Thanks for the contribution. I think this can be merged when the comments are addressed.

File: opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp
Comment (osae) [reply opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp:1236]: Problem: Call to 'solventWellRate' is used without being protected by 'has_solvent' (e.g. 'StandardWell').

Comment (totto82) [reply opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp:1236]: I see. 
I suggest either adding if(!has_sovlent) return 0.0 in solventWellRate(...) or guards usage of solventWellRate with if(has_solvent)

File: opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp
Comment (totto82) [inline opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp:624]: why is wt.second[1] removed?

**Comment by totto82** [comment]
Looks good. I have one minor comment.

**Comment by totto82** [comment]
jenkins build this opm-common=2012  opm-models=628 please

**Comment by totto82** [comment]
You need to add opm/models/blackoil/blackoilextbomodules.hh to the CMakeLists_files.cmake in opm-models

**Comment by osae** [comment]
jenkins build this opm-common=2012 opm-models=628 please

**Comment by totto82** [comment]
jenkins build this opm-common=2012 opm-models=628 please

**Comment by osae** [comment]
jenkins build this opm-common=2012 opm-models=628 please

**Comment by bska** [comment]
jenkins build this opm-common=2012 opm-models=628 please

**Comment by osae** [comment]
jenkins build this opm-common=2012 opm-models=628 please

**Comment by totto82** [comment]
jenkins build this opm-common=2012 opm-models=628 please

**Comment by osae** [comment]
jenkins build this opm-common=2012 opm-models=628 please

**Comment by totto82** [comment]
jenkins build this opm-common=2012 opm-models=628 please

File: opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp
Comment (osae) [reply opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp]: This was not intended. Now fixed.

File: opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp
Comment (osae) [reply opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp]: Ok.

File: ebos/eclfluxmodule.hh
Comment (osae) [reply ebos/eclfluxmodule.hh:278]: Ok

File: ebos/ecloutputblackoilmodule.hh
Comment (osae) [reply ebos/ecloutputblackoilmodule.hh]: Ok

File: ebos/ecloutputblackoilmodule.hh
Comment (osae) [reply ebos/ecloutputblackoilmodule.hh:1091]: Ok

**Comment by totto82** [comment]
jenkins build this opm-common=2012 opm-models=628 please

**Comment by totto82** [comment]
We are closing in. There is some new warning for unused variable. See comment in opm-models. These should be fixed before merging.

**Comment by totto82** [comment]
jenkins build this opm-common=2012 opm-models=628 please

**Comment by totto82** [comment]
Thanks @osae Lets merge it!

**Pull request merged**
